### ISSUE TITLE
Login with OpenIdConnect gives FOREIGN KEY constraint failed error

### ISSUE THREAD

**Issue by jbreuer** [umbraco/Umbraco-CMS#12853]
### Which *exact* Umbraco version are you using? For example: 9.0.1 - don't just write v9

10.1.0

### Bug summary

I was trying to do the same setup as from https://github.com/umbraco/Umbraco-CMS/issues/12749, but on a new 10.1.0 project with SQLite. When I try to login I'm redirect to the Identity Provider. After I login there I'm redirected back to Umbraco, but that gives the following error:

`SQLite Error 19: 'FOREIGN KEY constraint failed'.`

![image](https://user-images.githubusercontent.com/5444597/184599533-4d279cc0-197d-4ae4-b553-6c125d314508.png)

### Specifics

This is the full error from the logs:

```
Microsoft.Data.Sqlite.SqliteException (0x80004005): SQLite Error 19: 'FOREIGN KEY constraint failed'.
   at Microsoft.Data.Sqlite.SqliteException.ThrowExceptionForRC(Int32 rc, sqlite3 db)
   at Microsoft.Data.Sqlite.SqliteDataReader.NextResult()
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader(CommandBehavior behavior)
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader()
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteNonQuery()
   at Umbraco.Cms.Persistence.Sqlite.Services.SqlitePreferDeferredTransactionsConnection.CommandWrapper.ExecuteNonQuery()
   at StackExchange.Profiling.Data.ProfiledDbCommand.ExecuteNonQuery() in C:\projects\dotnet\src\MiniProfiler.Shared\Data\ProfiledDbCommand.cs:line 281
   at Umbraco.Cms.Infrastructure.Persistence.FaultHandling.FaultHandlingDbCommand.<ExecuteNonQuery>b__32_0()
   at Umbraco.Cms.Infrastructure.Persistence.FaultHandling.FaultHandlingDbCommand.<>c__DisplayClass38_0`1.<Execute>b__0()
   at Umbraco.Cms.Infrastructure.Persistence.FaultHandling.RetryPolicy.ExecuteAction[TResult](Func`1 func)
   at Umbraco.Cms.Infrastructure.Persistence.FaultHandling.FaultHandlingDbCommand.Execute[T](Func`1 f)
   at Umbraco.Cms.Infrastructure.Persistence.FaultHandling.FaultHandlingDbCommand.ExecuteNonQuery()
   at NPoco.Database.<>c__DisplayClass296_0.<ExecuteNonQueryHelper>b__0()
   at NPoco.Database.ExecutionHook[T](Func`1 action)
   at NPoco.Database.ExecuteNonQueryHelper(DbCommand cmd)
   at NPoco.Database.InsertAsyncImp[T](PocoData pocoData, String tableName, String primaryKeyName, Boolean autoIncrement, T poco, Boolean sync)
   at NPoco.AsyncHelper.RunSync[T](Task`1 task)
   at NPoco.Database.Insert[T](String tableName, String primaryKeyName, Boolean autoIncrement, T poco)
   at NPoco.Database.Insert[T](T poco)
   at NPoco.DatabaseType.InsertBulk[T](IDatabase db, IEnumerable`1 pocos, InsertBulkOptions options)
   at NPoco.Database.InsertBulk[T](IEnumerable`1 pocos, InsertBulkOptions options)
```
Timestamp	2022-08-15T07:59:30.3014175+00:00
@MessageTemplate	Exception ({InstanceId}).
InstanceId	7865c0d9
SourceContext	Umbraco.Cms.Infrastructure.Persistence.UmbracoDatabase 
ActionId	4351ce4e-1744-44b1-9368-d4ab995b85e4
ActionName	Umbraco.Cms.Web.Website.Controllers.UmbExternalLoginController.ExternalLoginCallback (Umbraco.Web.Website)
RequestId	800000c3-0002-fa00-b63f-84710c7967bb
RequestPath	/login/
ProcessId	9964
ProcessName	iisexpress
ThreadId	66
ApplicationId	0521d4839bc6f50f6ff21829f41b9aeff4fb79a6
MachineName	4QG5RQ2 
Log4NetLevel	ERROR
HttpRequestId	4803ef19-5686-4c97-8e34-aa5fd4f36718
HttpRequestNumber	4 
HttpSessionId	ea74524a-ef0f-e05a-ebc4-35f00479889d

### Steps to reproduce

Follow the steps to reproduce from https://github.com/umbraco/Umbraco-CMS/issues/12749 on a 10.1.0 site with SQLite. Then try to login to the external provider with a new member.

### Expected result / actual result

After I login into the external provider I should also be logged into Umbraco. My updated tokens should be stored in the umbracoExternalLoginToken table.

---
_This item has been added to our backlog [AB#21855](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/21855)_

**Comment by nul800sebastiaan** [comment]
I'll add this info to the original bug so they can look at it at the same time.

**Comment by nul800sebastiaan** [comment]
No actually, this should be a new issue, it's probably a SQLite problem, I would recommend trying on LocalDb / SQL Server instead for now.

**Comment by jbreuer** [comment]
I'm working on an example project which shows how OpenIdConnect can be used with Umbraco (still WIP so no readme yet): https://github.com/jbreuer/Umbraco-OpenIdConnect-Example

It comes with SQLite so people only need to download it and can just run the project. No need to change any configuration. That's why I would like this to also work with SQLite.

**Comment by nul800sebastiaan** [comment]
I understand that reminds me: I am also a bit scared that people will then start building on a clone of that example and when it comes to deploying the project they'll realize that SQLite is not a feasible option for larger sites in production. And if you need OpenIdConnect then it will very likely be a larger site. 

But it's a good one to get fixed in any case, just want to make sure that you are aware of that.

**Comment by jbreuer** [comment]
> I understand that reminds me: I am also a bit scared that people will then start building on a clone of that example and when it comes to deploying the project they'll realize that SQLite is not a feasible option for larger sites in production. And if you need OpenIdConnect then it will very likely be a larger site.

That's a valid point. The example is only so people can see how all the OpenIdConnect parts and Umbraco work together. The demo site should not be used for production. I'll add that to the docs.

**Comment by jbreuer** [comment]
The example project can already be used to reproduce the error. If you download the latest version and change the OpenIdConnect settings you can run the project: https://github.com/jbreuer/Umbraco-OpenIdConnect-Example/blob/main/Umbraco-OpenIdConnect-Example.Web/appsettings.json#L30

Than if you login you will get the error:
![image](https://user-images.githubusercontent.com/5444597/185119638-0e073b6d-f43b-4f24-aabb-37694268af66.png)

**Comment by jbreuer** [comment]
I've added the ClientId and ClientSecret to the repo. So you can just download the project and run it. Somehow if you press the `Sign in with OpenIdConnect` button it takes about 1 min the first time. Not sure why, but now you have a working example to reproduce the error.

Normally the ClientSecret should not be in the Github repo.
These settings are valid and only used for this example.
So it's ok these are public.

**Comment by jbreuer** [comment]
I changed the db from the example project from SQLite to SQL Server. Than it works:
![image](https://user-images.githubusercontent.com/5444597/185174459-c5b8b758-0d38-485a-9e48-867a5f6895f0.png)
So it's really a SQLite problem.

**Comment by jbreuer** [comment]
I used SourceLink to debug the Umbraco source code. I found out what line is causing the error.

After login the following method is called:

`Umbraco.Cms.Web.Website.Controllers.UmbExternalLoginController.ExternalLoginCallback`

A few steps deeper it will call this:

`Umbraco.Cms.Core.Security.MemberUserStore.CreateAsync`

And finally if a member group should be assigned it will call this:

`Umbraco.Cms.Infrastructure.Persistence.Repositories.Implement.MemberGroupRepository.AssignRolesInternal`

In that method the following code throws the error:

`IEnumerable<Member2MemberGroupDto> dtos = nonAssignedRoles.Select(x => new Member2MemberGroupDto { Member = mId, MemberGroup = rolesForNames[x!].NodeId });`

It's this line in the source code: https://github.com/umbraco/Umbraco-CMS/blob/v10/contrib/src/Umbraco.Infrastructure/Persistence/Repositories/Implement/MemberGroupRepository.cs#L277

So if you do an external login with a member that has a default member group this bug will happen in SQLite: https://github.com/jbreuer/Umbraco-OpenIdConnect-Example/blob/main/Umbraco-OpenIdConnect-Example.Core/Provider/OpenIdConnectMemberExternalLoginProviderOptions.cs#L41

**Comment by jbreuer** [comment]
I did some more research today and it's new bug that was introduced in V10. This commit made the changes that breaks it: https://github.com/umbraco/Umbraco-CMS/commit/f4e333c1782bcb7b614315db42a79702f67d328c

It's not just SQLite, but also SQL Server. I didn't get the error with SQL Server before because the member was already created. The problem is in this code:

Umbraco.Cms.Core.Security.MemberUserStore.CreateAsync
```
// create member
IMember memberEntity = _memberService.CreateMember(
    user.UserName,
    user.Email,
    user.Name.IsNullOrWhiteSpace() ? user.UserName : user.Name!,
    user.MemberTypeAlias.IsNullOrWhiteSpace()
        ? Constants.Security.DefaultMemberTypeAlias
        : user.MemberTypeAlias!);

UpdateMemberProperties(memberEntity, user);

// create the member
_memberService.Save(memberEntity);

// We need to add roles now that the member has an Id. It do not work implicit in UpdateMemberProperties
_memberService.AssignRoles(
    new[] { memberEntity.Id },
    user.Roles.Select(x => x.RoleId).Where(x => x is not null).ToArray());
```
In UpdateMemberProperties there is no member id yet. But it has the following code:

```
if (identityUser.IsPropertyDirty(nameof(MemberIdentityUser.Roles)))
{
    anythingChanged = true;

    var identityUserRoles = identityUser.Roles.Select(x => x.RoleId).ToArray();
    _memberService.ReplaceRoles(new[] { member.Id }, identityUserRoles);
}
```
So it wants to call `ReplaceRoles`, but there is no member id yet. That causes the FOREIGN KEY constraint error. 
So this code already needs a member id. This could be done changing _memberService.CreateMember to _memberService.CreateMemberWithIdentity. But other parts might also need to change than because they assume there is no id yet.

**Comment by jbreuer** [comment]
I just realised that this bug is probably a bit bigger than I first thought. Since it's not just SQLite, but also SQL Server. If you login with a new external member in V10 which has defaultMemberGroups in the AutoLinkOptions you will get this error.

**Comment by jbreuer** [comment]
Somehow it doesn't always happen with SQL Server because it allows 0 in the cmsMember2MemberGroup table even though there are FK contraints. In SQLite that is not allowed.

**Comment by jbreuer** [comment]
I have created a PR which fixes this issue: https://github.com/umbraco/Umbraco-CMS/pull/12868.
`_memberService.ReplaceRoles` only needs to be executed if a member already exists in the database. 
When a member is created it will still get the correct roles because `_memberService.AssignRoles` will be called later.

**Comment by nikolajlauridsen** [comment]
Fixed in #12872 Thank you so much for all your work troubleshooting and even fixing the issue 🎉 💪
